ws-client/src/client.rs

```diff
@@ -400,34 +465,56 @@ async fn background_task(
 			}
 
 			Either::Left((Some(FrontToBack::Batch(batch)), _)) => {
-				log::trace!("[backend]: client prepares to send batch request: {:?}", batch);
-				if let Err(e) = sender.start_batch_request(batch, &mut manager).await {
+				log::trace!("[backend]: client prepares to send batch request: {:?}", batch.raw);
+				// NOTE(niklasad1): annoying allocation.
+				if let Err(send_back) = manager.insert_pending_batch(batch.ids.clone(), batch.send_back) {
+					log::warn!("[backend]: batch request: {:?} already pending", batch.ids);
+					let _ = send_back.send(Err(Error::InvalidRequestId));
+					continue;
+				}
+
+				if let Err(e) = sender.send(batch.raw).await {
 					log::warn!("[backend]: client batch request failed: {:?}", e);
+					manager.complete_pending_batch(batch.ids);
 				}
 			}
-
 			// User called `notification` on the front-end
 			Either::Left((Some(FrontToBack::Notification(notif)), _)) => {
 				log::trace!("[backend]: client prepares to send notification: {:?}", notif);
-				if let Err(e) = sender.send_notification(notif).await {
+				if let Err(e) = sender.send(notif).await {
 					log::warn!("[backend]: client notif failed: {:?}", e);
 				}
 			}
 
 			// User called `request` on the front-end
-			Either::Left((Some(FrontToBack::StartRequest(request)), _)) => {
+			Either::Left((Some(FrontToBack::Request(request)), _)) => {
 				log::trace!("[backend]: client prepares to send request={:?}", request);
-				if let Err(e) = sender.start_request(request, &mut manager).await {
-					log::warn!("[backend]: client request failed: {:?}", e);
+				match sender.send(request.raw).await {
+					Ok(_) => manager
+						.insert_pending_call(request.id, request.send_back)
+						.expect("ID unused checked above; qed"),
+					Err(e) => {
+						log::warn!("[backend]: client request failed: {:?}", e);
+						let _ = request.send_back.map(|s| s.send(Err(Error::TransportError(Box::new(e)))));
+					}
 				}
 			}
+
 			// User called `subscribe` on the front-end.
-			Either::Left((Some(FrontToBack::Subscribe(subscribe)), _)) => {
-				log::trace!("[backend]: client prepares to start subscription: {:?}", subscribe);
-				if let Err(e) = sender.start_subscription(subscribe, &mut manager).await {
+			Either::Left((Some(FrontToBack::Subscribe(sub)), _)) => match sender.send(sub.raw).await {
+				Ok(_) => manager
+					.insert_pending_subscription(
+						sub.subscribe_id,
+						sub.unsubscribe_id,
```

Comment:
to be on the safe-side we should probably register the unsubscribe ID in request manager, such that it can't be used by another request concurrently such that the unsubscription request gets thrown away.

technically, that's very unlikely to happen but still the counter can wrap around re-use request IDs,